### ISSUE TITLE
Runtime Error when using ExclamationTopology example in mini cluster (1 nimbus, 2 workers) using HDFS and Aurora under Ubuntu 14.04

### ISSUE THREAD

**Issue by lilalinda** [twitter/heron#1027]
My setup is the same as described in [Supun's blog](http://streamanalytics.blogspot.de/2016/06/deploying-heron-on-cluster-of-machines.html): 3 machines running Ubuntu 14.04 (server) on a local network 10.9.9.x with one designated as the master (nimbus) and two workers (supervisor1 and supervisor2).  Not using Vagrant and/or Docker at the moment.  I am trying to run the ExclamationTopology example with Heron 0.14.0 but keep getting a C stack trace that I cannot make sense of.  Interestingly, only the first worker task being dispatched shows the stack trace involving Google protobuf while subsequent tasks (on both workers) then show failures in `epoll_wait`.

I am submitting the topology to Aurora with
```
$ heron submit --verbose aurora/hadoopuser/test ~/.heron/examples/heron-examples.jar com.twitter.heron.examples.ExclamationTopology ExclamationTopology
```
and have hacked the file `.heron/conf/aurora/heron.aurora` to use HDFS for the Aurora processes to distribute the Heron core and topology files:
```
  name = 'fetch_heron_system',
  cmdline = '/home/hadoopuser/hadoop/bin/hadoop --loglevel DEBUG fs -get %s %s && tar zxf %s' % (heron_core_release_uri, core_release_file, core_release_file)
```
and
```
  name = 'fetch_user_package',
  cmdline = '/home/hadoopuser/hadoop/bin/hadoop fs -get %s %s && tar zxf %s' % (heron_topology_jar_uri, topology_package_file, topology_package_file)
```

Then, it seems that the two files get successfully fetched but the first time the process `launch_heron_executor` is run on one of the workers, it fails like this (STDERR) in some code relating to Google protobuf:
![worker2-executor-stderr](https://cloud.githubusercontent.com/assets/1231235/16577672/dcf534d8-4296-11e6-9b48-ebad8136625b.png)
The command line was:
![worker2-executor-cmdline](https://cloud.githubusercontent.com/assets/1231235/16577686/ec8f5b26-4296-11e6-9b96-d5ebf5c6708a.png)

Subsequent times that the Aurora job gets dispatched, it still fails on the workers but this time with such a stack trace and always in `epoll_wait`:
![worker2-executor-stderr2](https://cloud.githubusercontent.com/assets/1231235/16577716/111f66fc-4297-11e6-8d2b-1b652ff7b72b.png)

If of any help, it seems to always take about 6-7 seconds for the executor to fail, i.e., the stack trace appears after the last regular log output.

I am pasting again the stack traces from the first time failure of three different runs, because the location is (unfortunately) different from run to run:
```
*** Aborted at 1467703906 (unix time) try "date -d @1467703906" if you are using GNU date ***
PC: @           0x4d37ce google::protobuf::io::CodedInputStream::PopLimit()
*** SIGTERM (@0x3e900006a17) received by PID 27173 (TID 0x7f404adcf780) from PID 27159; stack trace: ***
    @     0x7f404a3a6330 (unknown)
    @           0x4d37ce google::protobuf::io::CodedInputStream::PopLimit()
    @           0x466c5b heron::proto::system::HeronDataTupleSet::MergePartialFromCodedStream()
    @           0x467011 heron::proto::system::HeronTupleSet::MergePartialFromCodedStream()
    @           0x4560a1 heron::proto::stmgr::TupleMessage::MergePartialFromCodedStream()
    @           0x4d2287 google::protobuf::MessageLite::ParseFromArray()
    @           0x490b8d IncomingPacket::UnPackProtocolBuffer()
    @           0x42104b Server::dispatchMessage<>()
    @           0x492c45 Server::OnNewPacket()
    @           0x48cec9 Connection::handleDataRead()
    @           0x48a0dc BaseConnection::handleRead()
    @           0x48f94d EventLoopImpl::handleReadCallback()
    @           0x498fcc event_base_loop
    @           0x409416 main
    @     0x7f4049bcff45 __libc_start_main
    @           0x40d82f (unknown)
    @                0x0 (unknown)
```
```
*** Aborted at 1467640839 (unix time) try "date -d @1467640839" if you are using GNU date ***
PC: @           0x46722c heron::proto::system::HeronDataTupleSet::SerializeWithCachedSizesToArray()
*** SIGTERM (@0x3e900005fe3) received by PID 24562 (TID 0x7fed1941f780) from PID 24547; stack trace: ***
    @     0x7fed189f6330 (unknown)
    @           0x46722c heron::proto::system::HeronDataTupleSet::SerializeWithCachedSizesToArray()
    @           0x46731b heron::proto::system::HeronTupleSet::SerializeWithCachedSizesToArray()
    @           0x4564d6 heron::proto::stmgr::TupleMessage::SerializeWithCachedSizesToArray()
    @           0x4d2742 google::protobuf::MessageLite::SerializePartialToArray()
    @           0x490fec OutgoingPacket::PackProtocolBuffer()
    @           0x491f78 Server::SendResponse()
    @           0x49215e Server::SendMessage()
    @           0x41f3c9 heron::stmgr::StMgrServer::SendToInstance()
    @           0x415d54 heron::stmgr::StMgr::SendInBound()
    @           0x415fcb heron::stmgr::StMgr::DrainInstanceData()
    @           0x42ac2f heron::stmgr::TupleCache::TupleList::drain()
    @           0x42ad0a heron::stmgr::TupleCache::drain_impl()
    @           0x490305 EventLoopImpl::handleTimerCallback()
    @           0x48e0db EventLoopImpl::eventLoopImplTimerCallback()
    @           0x498fcc event_base_loop
    @           0x409416 main
    @     0x7fed1821ff45 __libc_start_main
    @           0x40d82f (unknown)
    @                0x0 (unknown)
```
```
*** Aborted at 1467641426 (unix time) try "date -d @1467641426" if you are using GNU date ***
PC: @           0x46381f google::protobuf::internal::RepeatedPtrFieldBase::Destroy<>()
*** SIGTERM (@0x3e900006331) received by PID 25408 (TID 0x7f8f91293780) from PID 25393; stack trace: ***
    @     0x7f8f9086a330 (unknown)
    @           0x46381f google::protobuf::internal::RepeatedPtrFieldBase::Destroy<>()
    @           0x4638cb heron::proto::system::HeronDataTuple::~HeronDataTuple()
    @           0x463901 heron::proto::system::HeronDataTuple::~HeronDataTuple()
    @           0x463942 google::protobuf::internal::RepeatedPtrFieldBase::Destroy<>()
    @           0x4639b8 heron::proto::system::HeronDataTupleSet::~HeronDataTupleSet()
    @           0x4639e1 heron::proto::system::HeronDataTupleSet::~HeronDataTupleSet()
    @           0x466004 heron::proto::system::HeronTupleSet::SharedDtor()
    @           0x466048 heron::proto::system::HeronTupleSet::~HeronTupleSet()
    @           0x466071 heron::proto::system::HeronTupleSet::~HeronTupleSet()
    @           0x450abb heron::proto::stmgr::TupleMessage::~TupleMessage()
    @           0x450af1 heron::proto::stmgr::TupleMessage::~TupleMessage()
    @           0x423faf heron::stmgr::StMgrServer::HandleTupleSetMessage()
    @           0x4210ad Server::dispatchMessage<>()
    @           0x492c45 Server::OnNewPacket()
    @           0x48cec9 Connection::handleDataRead()
    @           0x48a0dc BaseConnection::handleRead()
    @           0x48f94d EventLoopImpl::handleReadCallback()
    @           0x498fcc event_base_loop
    @           0x409416 main
    @     0x7f8f90093f45 __libc_start_main
    @           0x40d82f (unknown)
    @                0x0 (unknown)
```

Any thoughts are appreciated!

**Comment by kramasamy** [comment]
@lilalinda - I see an issue in the command line 

```
$ heron submit --verbose aurora/hadoopuser/test ~/.heron/examples/heron-examples.jar com.twitter.heron.examples.ExclamationTopology ExclamationTopology
```

**Comment by maosongfu** [comment]
@lilalinda 
Can u share the log files in failed container's sandbox? They are:
-- heron-executor.stderr
-- heron-executor.stdout
-- all files in log-files folder

**Comment by lilalinda** [comment]
@kramasamy If I leave out role and environment (i.e., `/hadoopuser/test`) in the cluster specification, it tells me:
```
hadoopuser@nimbus:~$ heron submit --verbose aurora ~/.heron/examples/heron-examples.jar com.twitter.heron.examples.ExclamationTopology ExclamationTopology      ERROR: Argument cluster/[role]/[env] is not correct: role required but not provided
```
But I do see in my config files (using the Aurora ones from the release with minimal changes): 
```
hadoopuser@nimbus:~$ more .heron/conf/aurora/client.yaml
# location of the core package
#heron.package.core.uri:                      "file:///vagrant/.herondata/dist/heron-core-release.tar.gz"
heron.package.core.uri:                      "hdfs:///heron/topologies/heron-core.tar.gz"

# Whether role/env is required to submit a topology. Default value is False.
heron.config.is.role.required:               True
heron.config.is.env.required:                True
```
Now turning these off (by commenting them out) sets the role for the Aurora call to `hadoopuser` and the environment to `default` but then Aurora balks during the submit attempt with:
```
Error loading configuration: Environment should be one of "prod", "devel", "test" or staging<number>!  Got default
```
so I guess it is safer to keep using `test` or one of the other expected values.

**Comment by kramasamy** [comment]
@lilalinda - sure it is fine to keep the configuration and pass cluster/role/env - triplet. As @maosongfu pointed out, can you share the log files?

**Comment by kramasamy** [comment]
@lilalinda - just wondering if you have installed libunwind in both the master and the slave nodes.

**Comment by lilalinda** [comment]
@kramasamy I am trying to package the log files and think I have a useful bundle now -- will share in a minute.  (Are you guys on Pacific Time?  I am currently working from Germany...)

Yes, I have on all 3 nodes:
```
hadoopuser@nimbus:~$ dpkg -l | grep libunwind
ii  libunwind-setjmp0                   1.1-2.2ubuntu3                      amd64        libunwind-based non local goto - runtime
ii  libunwind-setjmp0-dev               1.1-2.2ubuntu3                      amd64        libunwind-based non local goto - development
ii  libunwind8                          1.1-2.2ubuntu3                      amd64        library to determine the call-chain of a program - runtime
ii  libunwind8-dev                      1.1-2.2ubuntu3                      amd64        library to determine the call-chain of a program - development
```

**Comment by kramasamy** [comment]
That is great to know libunwind is there. Sure, look forward to the logs. We are in Pacific Time.

**Comment by lilalinda** [comment]
@maosongfu and @kramasamy Here are the log files.  I created them on the 2nd slave node changing into the directory
`/var/lib/mesos/slaves/<HASH>/frameworks/<HASH>/executors/thermos-<AURORA-TASK>/runs/latest/`
and then:
```
$ tar -cvzf ~/test-logs.tgz --exclude=*.tar.gz --exclude=*.jar --exclude=*.pex --exclude=.pex --exclude=heron-core .

```
Finally, I had to compress it again with 7z to .zip format for github to allow this file type to be uploaded.
(Let me know if there is a better place to grab the logs.
[test-logs.zip](https://github.com/twitter/heron/files/349301/test-logs.zip)
)

**Comment by kramasamy** [comment]
@ajorgensen - can you help by looking at the logs - what could be going wrong? you are the closest since your environment is similar to @lilalinda

**Comment by ajorgensen** [comment]
Can you check if core dumps are turned on `ulimit -c`. If that is set to 0 you can turn it on and re-run the job to see if it will give us some more information about what is going on. Here's an article on how to turn it on for everyone `http://www.akadia.com/services/ora_enable_core.html`. You can also see where Ubuntu will store the core dumps by looking at `cat /proc/sys/kernel/core_pattern` which on my machines is set to `/proc/%p/cwd/core.%p` which should put the core dump in the root directory of the sandbox of the job.

We generally have core dumps turned off as they have the potential to fill up the disk really fast depending on the failure and only turn them on when necessary to gather more information. It's probably also possible to run it with a limit thats not unlimited to help curb this but I've generally found we only need core dumps in very rare cases.

**Comment by maosongfu** [comment]
@lilalinda 
By design, heron-executor spawns children processes and restart them if any failure happens unless the amount of retry exceeds the limitation.

According to the log in stderr:
```
*** Aborted at 1467703906 (unix time) try "date -d @1467703906" if you are using GNU date ***
PC: @           0x4d37ce google::protobuf::io::CodedInputStream::PopLimit()
*** SIGTERM (@0x3e900006a17) received by PID 27173 (TID 0x7f404adcf780) from PID 27159; stack trace: ***
    @     0x7f404a3a6330 (unknown)
    @           0x4d37ce google::protobuf::io::CodedInputStream::PopLimit()
    @           0x466c5b heron::proto::system::HeronDataTupleSet::MergePartialFromCodedStream()
    @           0x467011 heron::proto::system::HeronTupleSet::MergePartialFromCodedStream()
    @           0x4560a1 heron::proto::stmgr::TupleMessage::MergePartialFromCodedStream()
    @           0x4d2287 google::protobuf::MessageLite::ParseFromArray()
    @           0x490b8d IncomingPacket::UnPackProtocolBuffer()
    @           0x42104b Server::dispatchMessage<>()
    @           0x492c45 Server::OnNewPacket()
    @           0x48cec9 Connection::handleDataRead()
    @           0x48a0dc BaseConnection::handleRead()
    @           0x48f94d EventLoopImpl::handleReadCallback()
    @           0x498fcc event_base_loop
    @           0x409416 main
    @     0x7f4049bcff45 __libc_start_main
    @           0x40d82f (unknown)
    @                0x0 (unknown)
```
The stream mgr had some troubles.

According to the log in heron-executor.stdout:
```
2016-07-05 09:31:39: Logging pid 27173 to file stmgr-1.pid
2016-07-05 09:31:39: Running metricsmgr-1 process as /usr/lib/jvm/java-8-oracle/bin/java -Xmx1024M -XX:+PrintCommandLineFlags -verbosegc -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -XX:+PrintGCCause -XX:+PrintPromotionFailure -XX:+PrintTenuringDistribution -XX:+PrintHeapAtGC -XX:+HeapDumpOnOutOfMemoryError -XX:+UseConcMarkSweepGC -XX:+PrintCommandLineFlags -Xloggc:log-files/gc.metricsmgr.log -Djava.net.preferIPv4Stack=true -cp ./heron-core/lib/metricsmgr/* com.twitter.heron.metricsmgr.MetricsManager metricsmgr-1 31288 ExclamationTopology ExclamationTopology8c1920b3-e596-4af1-a626-d2457c5f9689 ./heron-conf/heron_internals.yaml ./heron-conf/metrics_sinks.yaml
2016-07-05 09:31:39: Logging pid 27174 to file metricsmgr-1.pid
2016-07-05 09:31:46: Executor terminated; exiting all process in executor.
```

It seems the heron-executor exited directly without restarting the failed stream mgr.
And I found in the thermos_runner.INFO:
```
I0705 09:31:38.587270 27058 runner.py:845] Forking Process(launch_heron_executor)
I0705 09:31:46.608158 27058 thermos_runner.py:160] Thermos runner got signal 10, shutting down.
I0705 09:31:46.608392 27058 thermos_runner.py:161] Interrupted frame:
I0705 09:31:46.610755 27058 thermos_runner.py:164]   File "/usr/lib/python2.7/runpy.py", line 162, in _run_module_as_main
I0705 09:31:46.610853 27058 thermos_runner.py:164]     "__main__", fname, loader, pkg_name)
I0705 09:31:46.610932 27058 thermos_runner.py:164]   File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
I0705 09:31:46.611006 27058 thermos_runner.py:164]     exec code in run_globals
I0705 09:31:46.611077 27058 thermos_runner.py:164]   File "/var/lib/mesos/slaves/a12ea196-4416-4904-b123-e26b2ba82f22-S3/frameworks/1e37e686-4cc7-465d-bc0c-a2b6e1f286f4-0000/executors/thermos-hadoopuser-test-ExclamationTopology-1-961438f4-a36c-4a66-94d9-d4ce0ad443b5/runs/69b97000-1547-4262-9be9-9ece49c254a5/thermos_runner.pex/__main__.py", line 25, in <module>
I0705 09:31:46.611149 27058 thermos_runner.py:164]     bootstrap_pex(__entry_point__)
I0705 09:31:46.611217 27058 thermos_runner.py:164]   File "/var/lib/mesos/slaves/a12ea196-4416-4904-b123-e26b2ba82f22-S3/frameworks/1e37e686-4cc7-465d-bc0c-a2b6e1f286f4-0000/executors/thermos-hadoopuser-test-ExclamationTopology-1-961438f4-a36c-4a66-94d9-d4ce0ad443b5/runs/69b97000-1547-4262-9be9-9ece49c254a5/thermos_runner.pex/.bootstrap/_pex/pex_bootstrapper.py", line 98, in bootstrap_pex
I0705 09:31:46.611290 27058 thermos_runner.py:164]     pex.PEX(entry_point).execute()
I0705 09:31:46.611367 27058 thermos_runner.py:164]   File "/var/lib/mesos/slaves/a12ea196-4416-4904-b123-e26b2ba82f22-S3/frameworks/1e37e686-4cc7-465d-bc0c-a2b6e1f286f4-0000/executors/thermos-hadoopuser-test-ExclamationTopology-1-961438f4-a36c-4a66-94d9-d4ce0ad443b5/runs/69b97000-1547-4262-9be9-9ece49c254a5/thermos_runner.pex/.bootstrap/_pex/pex.py", line 324, in execute
I0705 09:31:46.611448 27058 thermos_runner.py:164]     self._wrap_coverage(self._wrap_profiling, self._execute)
I0705 09:31:46.611558 27058 thermos_runner.py:164]   File "/var/lib/mesos/slaves/a12ea196-4416-4904-b123-e26b2ba82f22-S3/frameworks/1e37e686-4cc7-465d-bc0c-a2b6e1f286f4-0000/executors/thermos-hadoopuser-test-ExclamationTopology-1-961438f4-a36c-4a66-94d9-d4ce0ad443b5/runs/69b97000-1547-4262-9be9-9ece49c254a5/thermos_runner.pex/.bootstrap/_pex/pex.py", line 257, in _wrap_coverage
I0705 09:31:46.611659 27058 thermos_runner.py:164]     runner(*args)
I0705 09:31:46.611751 27058 thermos_runner.py:164]   File "/var/lib/mesos/slaves/a12ea196-4416-4904-b123-e26b2ba82f22-S3/frameworks/1e37e686-4cc7-465d-bc0c-a2b6e1f286f4-0000/executors/thermos-hadoopuser-test-ExclamationTopology-1-961438f4-a36c-4a66-94d9-d4ce0ad443b5/runs/69b97000-1547-4262-9be9-9ece49c254a5/thermos_runner.pex/.bootstrap/_pex/pex.py", line 289, in _wrap_profiling
I0705 09:31:46.611821 27058 thermos_runner.py:164]     runner(*args)
I0705 09:31:46.611890 27058 thermos_runner.py:164]   File "/var/lib/mesos/slaves/a12ea196-4416-4904-b123-e26b2ba82f22-S3/frameworks/1e37e686-4cc7-465d-bc0c-a2b6e1f286f4-0000/executors/thermos-hadoopuser-test-ExclamationTopology-1-961438f4-a36c-4a66-94d9-d4ce0ad443b5/runs/69b97000-1547-4262-9be9-9ece49c254a5/thermos_runner.pex/.bootstrap/_pex/pex.py", line 367, in _execute
I0705 09:31:46.611958 27058 thermos_runner.py:164]     return self.execute_entry(self._pex_info.entry_point)
I0705 09:31:46.612025 27058 thermos_runner.py:164]   File "/var/lib/mesos/slaves/a12ea196-4416-4904-b123-e26b2ba82f22-S3/frameworks/1e37e686-4cc7-465d-bc0c-a2b6e1f286f4-0000/executors/thermos-hadoopuser-test-ExclamationTopology-1-961438f4-a36c-4a66-94d9-d4ce0ad443b5/runs/69b97000-1547-4262-9be9-9ece49c254a5/thermos_runner.pex/.bootstrap/_pex/pex.py", line 425, in execute_entry
I0705 09:31:46.612095 27058 thermos_runner.py:164]     runner(entry_point)
I0705 09:31:46.612162 27058 thermos_runner.py:164]   File "/var/lib/mesos/slaves/a12ea196-4416-4904-b123-e26b2ba82f22-S3/frameworks/1e37e686-4cc7-465d-bc0c-a2b6e1f286f4-0000/executors/thermos-hadoopuser-test-ExclamationTopology-1-961438f4-a36c-4a66-94d9-d4ce0ad443b5/runs/69b97000-1547-4262-9be9-9ece49c254a5/thermos_runner.pex/.bootstrap/_pex/pex.py", line 430, in execute_module
I0705 09:31:46.612236 27058 thermos_runner.py:164]     runpy.run_module(module_name, run_name='__main__')
I0705 09:31:46.612348 27058 thermos_runner.py:164]   File "/usr/lib/python2.7/runpy.py", line 180, in run_module
I0705 09:31:46.612421 27058 thermos_runner.py:164]     fname, loader, pkg_name)
I0705 09:31:46.612488 27058 thermos_runner.py:164]   File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
I0705 09:31:46.612555 27058 thermos_runner.py:164]     exec code in run_globals
```

So there was something happened at ~ 09:31:46.
So I have a question:
1. Did u run "heron kill ..." to kill the job at ~ 09:31:46 ?
2. Could u check whether the job hit aurora limitation (for instance, used more ram than required) and was killed by aurora?

Also, currently the stderr & stdout of stream mgr can lose: (https://github.com/twitter/heron/issues/958)
So can u try to change here: https://github.com/twitter/heron/blob/master/heron/executor/src/python/heron-executor.py#L359
to forcibly redirect the logs to a file and share the content of the file with us?

Thanks

**Comment by wuye9036** [comment]
In POSIX standard, it means SIGUSR1.

https://github.com/torvalds/linux/blob/master/include/uapi/asm-generic/signal.h

Which is same as POSIX.

Aurora treats this SIGNAL same as KILL.

And could we get the log from slave.cpp of aurora?

I think we could find more details in it.

**Comment by maosongfu** [comment]
@wuye9036 Aurora is open sourced: https://github.com/apache/aurora

**Comment by lilalinda** [comment]
@maosongfu Re your 3 questions:
1. No, I did not run the "heron kill... " command at that time.  I ran it a couple minutes later when the subsequently launched tasks also failed.  The logs you are referring to are the ones from the first task incarnation.
2. @wuye9036 commented on this, I believe.
3. I did not check any Aurora specific logs but you were right: it seems to be a RAM issue!

I followed your guidance and re-built Heron from the latest source code (git `4cbe5ae96b72f`) and it generated a Java Heap out-of-memory error!  The two slave nodes are each configured with 4GB of RAM, and Mesos advertised about 2.7GB available on each.  When I ran the ExclamationTopology out of the box (the code requests 512MB for each of the 2 containers: word and exclaim, so I though it should be around 1GB), and looked at the --verbose output of "heron submit --verbose ..." I saw that Aurora blew it up to 3GB RAM_PER_CONTAINER!

Next, I tried lowering the RAM requested in the source code to 1MB (failed immediately upon "heron submit" asking for a minimum of 192MB) then to 200MB so that it would have a chance to run on my slave nodes.  Then I encountered the Java Heap Space error, which would explain how it failed at different points in the Google Protobuf code (because the out-of-memory gets hit at different times from run to run).

Finally, I bit the bullet and powered everything down (well, just the 2 slaves) and increased the VMs to have 16GB rather than 4GB.  Now things seemed to somewhat work with the code recompiled at the original level of requesting 512MB per container.  I even got some log lines with the bolt counting some emits!  However, the topology still failed and I am running out of time now to debug this further.  I fear I won't get to it until Monday...

Thanks for all your help so far and have a great weekend.  I'll be looking into this further on Monday morning German time.  Best!

**Comment by maosongfu** [comment]
@lilalinda Cool nice to know that. Let's dig it deeper when you get a chance. Similarly, we may need your log-files to figure out what happened.

**Comment by lilalinda** [comment]
Happy Monday, everyone.  Back to the office and my Heron cluster.  I updated from git this morning to use the latest sources and am now running `

Still some issues but more details in the logs.  It looks like the TMaster thread consistently dies.  But I have no clue, why.  Maybe you guys could take a look at the logs?  See below for a copy-and-paste and the attached ZIPs.

@ajorgensen I enabled core dumps but it looks like none got generated.  Well, since the Java Runtime Exceptions are now thrown, maybe this bypasses the core dumping in Ubuntu?  Here some details after my last run:
```
hadoopuser@supervisor1:~$ ulimit -a
core file size          (blocks, -c) unlimited
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 64103
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 64103
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
hadoopuser@supervisor1:~$ cat /proc/sys/kernel/core_pattern
|/usr/share/apport/apport %p %s %c %P
hadoopuser@supervisor1:~$ ls -lat /usr/share/apport/apport*
-rwxr-xr-x 1 root root 18241 May 18 11:44 /usr/share/apport/apport
-rwxr-xr-x 1 root root  1269 May 18 11:44 /usr/share/apport/apport-checkreports
-rwxr-xr-x 1 root root  3441 May 18 11:44 /usr/share/apport/apportcheckresume
-rw-r--r-- 1 root root  2366 May 18 11:44 /usr/share/apport/apport.jar
hadoopuser@supervisor1:~$ date
Mon Jul 11 11:27:59 CEST 2016
```

`supervisor1` shows this interesting output in `log-files/metricsmgr-0.log.0`:
```
[...]
[2016-07-11 11:24:13 +0200] com.twitter.heron.metricsmgr.MetricsManagerServer INFO: We received a new TMasterLocation. Register it into SingletonRegistry [2016-07-11 11:24:13 +0200] com.twitter.heron.metricsmgr.MetricsManagerServer INFO: Current TMaster location: topology_name: "ExclamationTopology" topology_id: "ExclamationTopology7a7981c0-c0d4-4d10-bd11-c5a2cecf7d70" host: "supervisor1" controller_port: 31945 master_port: 31118 stats_port: 31204 [2016-07-11 11:24:18 +0200] com.twitter.heron.metricsmgr.sink.tmaster.TMasterSink INFO: Update current TMasterLocation to: topology_name: "ExclamationTopology" topology_id: "ExclamationTopology7a7981c0-c0d4-4d10-bd11-c5a2cecf7d70" host: "supervisor1" controller_port: 31945 master_port: 31118 stats_port: 31204 [2016-07-11 11:24:18 +0200] com.twitter.heron.metricsmgr.sink.tmaster.TMasterSink SEVERE: Starting TMasterClient for the 1 time. [2016-07-11 11:24:18 +0200] com.twitter.heron.common.network.HeronClient INFO: Connecting to endpoint: supervisor1/127.0.1.1:31118 [2016-07-11 11:24:18 +0200] com.twitter.heron.metricsmgr.sink.tmaster.TMasterClient INFO: Connected to TMaster. Ready to send metrics [2016-07-11 11:24:20 +0200] com.twitter.heron.common.network.IncomingPacket SEVERE: channel.read returned negative -1 [2016-07-11 11:24:20 +0200] com.twitter.heron.common.network.SocketChannelHelper SEVERE: Something bad happened while reading from channel: /127.0.0.1:45008 [2016-07-11 11:24:20 +0200] com.twitter.heron.common.network.HeronServer INFO: Handling error from channel: /127.0.0.1:45008 [2016-07-11 11:24:20 +0200] com.twitter.heron.common.network.HeronServer INFO: Removing all interest on channel: /127.0.0.1:45008 [2016-07-11 11:24:20 +0200] com.twitter.heron.metricsmgr.MetricsManagerServer SEVERE: Got a connection close from remote socket address: /127.0.0.1:45008 [2016-07-11 11:24:20 +0200] com.twitter.heron.metricsmgr.MetricsManagerServer SEVERE: Un-register publish from hostname: supervisor1, component_name: __tmaster__, port: 31,118, instance_id: 0, instance_index: -1 [2016-07-11 11:24:20 +0200] com.twitter.heron.common.network.IncomingPacket SEVERE: channel.read returned negative -1 [2016-07-11 11:24:20 +0200] com.twitter.heron.common.network.SocketChannelHelper SEVERE: Something bad happened while reading from channel: supervisor1/127.0.1.1:31118 [2016-07-11 11:24:20 +0200] com.twitter.heron.common.network.HeronClient INFO: Handling Error. Cleaning states in HeronClient. [2016-07-11 11:24:20 +0200] com.twitter.heron.common.network.HeronClient INFO: Successfully closed the channel: java.nio.channels.SocketChannel[closed] [2016-07-11 11:24:20 +0200] com.twitter.heron.metricsmgr.sink.tmaster.TMasterClient SEVERE: Disconnected from TMaster. [2016-07-11 11:24:20 +0200] com.twitter.heron.metricsmgr.sink.tmaster.TMasterSink SEVERE: TMasterClient dies in thread: Thread[Thread-1,5,main] java.lang.RuntimeException: Errors happened due to write or read failure from TMaster. at com.twitter.heron.metricsmgr.sink.tmaster.TMasterClient.onError(TMasterClient.java:58) at com.twitter.heron.common.network.HeronClient.handleError(HeronClient.java:349) at com.twitter.heron.common.network.SocketChannelHelper.read(SocketChannelHelper.java:146) at com.twitter.heron.common.network.HeronClient.handleRead(HeronClient.java:162) at com.twitter.heron.common.basics.NIOLooper.handleSelectedKeys(NIOLooper.java:111) at com.twitter.heron.common.basics.NIOLooper.access$000(NIOLooper.java:32) at com.twitter.heron.common.basics.NIOLooper$1.run(NIOLooper.java:45) at com.twitter.heron.common.basics.WakeableLooper.executeTasksOnWakeup(WakeableLooper.java:142) at com.twitter.heron.common.basics.WakeableLooper.runOnce(WakeableLooper.java:74) at com.twitter.heron.common.basics.WakeableLooper.loop(WakeableLooper.java:64) at com.twitter.heron.metricsmgr.sink.tmaster.TMasterClient.run(TMasterClient.java:117) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) at java.lang.Thread.run(Thread.java:745) 
```
When I cd into the directory where the log files etc reside, namely `/var/lib/mesos/slaves/<HASH>/frameworks/<HASH>/executors/thermos-<AURORA-TASK>/runs/latest`, I see this interesting bit in the file `__main.log__`, which is not visible in the Aurora browsing (port 1338):
```
$ head -n 9 __main__.log
Log file created at: 2016/07/11 11:24:03
Running on machine: supervisor1
[DIWEF]mmdd hh:mm:ss.uuuuuu pid file:line] msg
Command line: /var/lib/mesos/slaves/a7058b40-59c4-4742-aa30-bd69cadb5ad1-S8/frameworks/1e37e686-4cc7-465d-bc0c-a2b6e1f286f4-0000/executors/thermos-hadoopuser-test-ExclamationTopology-0-941a7b9b-a46a-46e4-8581-61dd877659d8/runs/fe9a376b-f502-4714-9dea-eac9994a8087/thermos_executor.pex --announcer-ensemble 127.0.0.1:2181
I0711 11:24:03.502679 9436 executor_base.py:45] Executor [None]: registered() called with:
I0711 11:24:03.503627 9436 executor_base.py:45] Executor [None]:    ExecutorInfo:  executor_id {
  value: "thermos-hadoopuser-test-ExclamationTopology-0-941a7b9b-a46a-46e4-8581-61dd877659d8"
}
resources {
```
Here, the configuration with `--announcer-ensemble 127.0.0.1:2181` is what worries me:  The Zookeeper instance is NOT running on localhost but on `10.9.9.10:2181`, which is how I configured it in all possible places.  But could this be the culprit?  The Aurora processes `fetch_*` run through without a hitch but then the `launch_heron_executor` fails.  Could the mistaken reference to `--announcer-ensemble` cause the error later on???

Attached are the logs from the above cited run and generated with:
```
$ tar -cvzf ~/supervisor1-logs.tgz --exclude=*.tar.gz --exclude=*.jar --exclude=*.pex --exclude=.pex --exclude=heron-core .
$ cd
$ zip supervisor1-logs.zip supervisor1-logs.tgz
```
[supervisor1-logs.zip](https://github.com/twitter/heron/files/356829/supervisor1-logs.zip)

I am also looking at `supervisor2` and the same issues turn up there.

Finally, I am trying to read and understand the code at `heron/executor/src/python/heron-executor.py` and specifically how the class HeronExecutor gets initialized but I don't see an immediate place where the announcer ensemble gets conveyed.  Maybe it is some hidden Aurora configuration?  Which would be weird because I was able to run the Aurora hello_world example from their Getting Started tutorial issued from nimbus (master) to the supervisors (slaves) without a problem...

**Comment by lilalinda** [comment]
Finally, I got it to run!  It was indeed the `127.0.0.1:2181` culprit printed in `__main__.log`!

I wasn't able to backtrace where this setting came from but poking around I finally found the file `/etc/default/aurora-scheduler` on `nimbus` and changed the occurrences of 127.0.0.1 in the appropriate places such that:
```
$ grep "2181" /etc/default/aurora-scheduler
ZK_ENDPOINTS="10.9.9.10:2181"
THERMOS_EXECUTOR_FLAGS="--announcer-ensemble 10.9.9.10:2181"
```

Now I will roll back to the Heron 0.14.0 release and see if it works as well.  Closing this issue as it is solved.
